frame/dapps-staking/src/fix.rs

```diff
@@ -0,0 +1,175 @@
+                        staker
+                    );
+                    // every inner iteration also adds one read
+                    consumed_weight = consumed_weight.saturating_add(T::DbWeight::get().reads(1));
+                    let staked_value = staking_info.latest_staked_value();
+                    // get contract address as Vec<u8>
+                    let contract_address = PrefixIterator::last_raw_key(&contract_iter);
+                    // select staking_info for the contract
+                    let mut contract_stake_info = restake_fix
+                        .contract_staking_info
+                        .entry(contract_address.clone().to_vec())
+                        .or_default();
+
+                    // update total and number of stakers
+                    contract_stake_info.total =
+                        contract_stake_info.total.saturating_add(staked_value);
+                    contract_stake_info.number_of_stakers += 1;
+                    // check for total consumed weight
+                    // store last processed contract and break out if the limit is reached
+                    if consumed_weight >= weight_limit {
+                        log::info!(
+                            ">>> Ledger migration stopped after consuming {:?} weight.",
+                            consumed_weight
+                        );
+                        restake_fix.last_processed_staker_contract =
+                            Some(contract_address.clone().to_vec());
+                        break 'staker_iter;
+                    }
+                }
+
+                // when one staker is processed, store it and clear stored contract
+                restake_fix.last_processed_staker_contract = None;
+                restake_fix.last_fully_processed_staker =
+                    Some(Ledger::<T>::storage_map_final_key(staker));
+            }
+        } else {
+            log::info!(">>> Starting to empty the accumulator");
+            // if contractStakeInfo is empty, we're done
+            if restake_fix.contract_staking_info.is_empty() {
+                log::info!(">>> Accumulator empty, all done.");
+            } else {
+                // get current era and weight for that read
+                let current_era = Pallet::<T>::current_era();
+                consumed_weight = consumed_weight.saturating_add(T::DbWeight::get().reads(1));
+
+                let staking_info_iter = restake_fix.contract_staking_info.iter();
+
+                for (contract, contract_stake_info) in staking_info_iter {
+                    // write new ContractEraStake from corrected records in contract_staking_info
+                    // and remove that record afterwards
+                    // break out when the weight hits the limit
+                    ContractEraStake::<T>::insert(&contract, current_era, contract_stake_info);
+                    restake_fix.contract_staking_info.remove(contract);
```

Comment:
You shouldn't modify map while iterating over it.

Unfortunately, calls like `pop_first` are still experimental so we cannot use them.
And BTreeMap doesn't have a draining iterator.

But what you can do is create a copy of the key set and iterate over it.